packages/react-reconciler/src/ReactFiberBeginWork.js

```diff
@@ -202,67 +220,343 @@ export default function<T, P, I, TI, HI, PI, C, CC, CX, PL>(
+    const nextProps = workInProgress.pendingProps;
 
-    let shouldUpdate;
     if (current === null) {
-      if (!workInProgress.stateNode) {
-        // In the initial pass we might need to construct the instance.
-        constructClassInstance(workInProgress, workInProgress.pendingProps);
-        mountClassInstance(workInProgress, renderExpirationTime);
-        shouldUpdate = true;
+      if (workInProgress.stateNode === null) {
+        const ctor = workInProgress.type;
+        const unmaskedContext = getUnmaskedContext(workInProgress);
+        const needsContext = isContextConsumer(workInProgress);
+        const context = needsContext
+          ? getMaskedContext(workInProgress, unmaskedContext)
+          : emptyObject;
+        const instance = (workInProgress.stateNode = new ctor(
+          nextProps,
+          context,
+        ));
+        return mountClassComponent(
+          workInProgress,
+          instance,
+          nextProps,
+          hasContext,
+          needsContext,
+          unmaskedContext,
+          context,
+          renderExpirationTime,
+        );
       } else {
-        invariant(false, 'Resuming work not yet implemented.');
         // In a resume, we'll already have an instance we can reuse.
-        // shouldUpdate = resumeMountClassInstance(workInProgress, renderExpirationTime);
+        invariant(false, 'Resuming work not yet implemented.');
       }
     } else {
-      shouldUpdate = updateClassInstance(
+      return updateClassComponent(
         current,
         workInProgress,
+        hasContext,
         renderExpirationTime,
       );
     }
-    return finishClassComponent(
-      current,
+  }
+
+  function mountClassComponent(
+    workInProgress: Fiber,
+    instance: any,
+    nextProps: mixed,
+    hasContext: boolean,
+    needsContext: boolean,
+    unmaskedContext: any,
+    context: any,
+    renderExpirationTime: ExpirationTime,
+  ) {
+    instance.updater = classUpdater;
+    workInProgress.stateNode = instance;
+    // The instance needs access to the fiber so that it can schedule updates
+    ReactInstanceMap.set(instance, workInProgress);
+    if (__DEV__) {
+      instance._reactInternalInstance = fakeInternalInstance;
+    }
+
+    if (__DEV__) {
+      checkClassInstance(workInProgress);
+    }
+
+    let nextState = instance.state || null;
+
+    instance.props = nextProps;
+    instance.state = workInProgress.memoizedState = nextState;
+    instance.refs = emptyObject;
+    const nextContext = (instance.context = getMaskedContext(
+      workInProgress,
+      unmaskedContext,
+    ));
+
+    if (
+      enableAsyncSubtreeAPI &&
+      workInProgress.type != null &&
+      workInProgress.type.prototype != null &&
+      workInProgress.type.prototype.unstable_isAsyncReactComponent === true
+    ) {
+      workInProgress.internalContextTag |= AsyncUpdates;
+    }
+
+    if (typeof instance.componentWillMount === 'function') {
```

Comment:
For this to be wasm compatible, all of the instance access need to be separated. Including setting up props, state, and testing for existence of a life-cycle.

Reply:
@sebmarkbage Do we have any plan about wasm?